src/proto/single/ent.rs

```diff
@@ -74,6 +91,135 @@ impl JobBuilder {
+/// know state (e.g."enqueued", "working", "success", "unknown") and the date and time
+/// the job was last updated. Additionally, information on what's going on with the job
+/// ([desc](struct.ProgressUpdate.html#structfield.desc)) and completion percentage
+/// ([percent](struct.ProgressUpdate.html#structfield.percent)) may be available,
+/// if the worker provided those details.
+#[derive(Debug, Clone, Deserialize)]
+pub struct Progress {
+    /// Id of the tracked job.
+    pub jid: String,
+
+    /// Job's state.
+    pub state: String,
+
+    /// When this job was last updated.
+    #[serde(deserialize_with = "parse_datetime")]
+    pub updated_at: Option<DateTime<Utc>>,
+
+    /// Percentage of the job's completion.
+    pub percent: Option<u8>,
+
+    /// Arbitrary description that may be useful to whoever is tracking the job's progress.
+    pub desc: Option<String>,
+}
+
+/// Info on job execution progress (sent).
+///
+/// In Enterprise Faktory, a client executing a job can report on the execution
+/// progress, provided the job is trackable. A trackable job is the one with "track":1
+/// specified in the custom data hash.
+#[derive(Debug, Clone, Serialize, Builder)]
+#[builder(
+    custom_constructor,
+    setter(into),
+    build_fn(name = "try_build", private)
+)]
+pub struct ProgressUpdate {
+    /// Id of the tracked job.
+    #[builder(setter(custom))]
+    pub jid: String,
+
+    /// Percentage of the job's completion.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(default = "None")]
+    pub percent: Option<u8>,
+
+    /// Arbitrary description that may be useful to whoever is tracking the job's progress.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(default = "None")]
+    pub desc: Option<String>,
+
+    /// Allows to extend the job's reservation, if more time needed to execute it.
+    ///
+    /// Note that you cannot decrease the initial [reservation](struct.Job.html#structfield.reserve_for).
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(default = "None")]
+    pub reserve_until: Option<DateTime<Utc>>,
+}
+
+impl ProgressUpdate {
+    /// Create a new instance of `ProgressUpdateBuilder` with job ID already set.
+    ///
+    /// Equivalent to creating a [new](struct.ProgressUpdateBuilder.html#method.new)
+    /// `ProgressUpdateBuilder`.
+    pub fn builder(jid: impl Into<String>) -> ProgressUpdateBuilder {
+        ProgressUpdateBuilder {
+            jid: Some(jid.into()),
+            ..ProgressUpdateBuilder::create_empty()
+        }
+    }
+
+    /// Create a new instance of `ProgressUpdate`.
+    ///
+    /// While job ID is specified at `ProgressUpdate`'s creation time,
+    /// the rest of the [fields](struct.ProgressUpdate.html) are defaulted to _None_.
+    pub fn new(jid: impl Into<String>) -> ProgressUpdate {
+        ProgressUpdateBuilder::new(jid).build()
+    }
```

Comment:
Why is this helper useful? I feel like we should either make people always use the builder _or_ have the fields be `pub` and settable after the fact, not both.

May be useful to have a shortcut like this though:
```rust
pub fn set_progress(jid: impl Into<String>), percent: u8) -> ProgressUpdate
```

Reply:
Indeed, progress update without an actual update does not make any sense (apart from the fact that the worker _did_  send an update, like a heart-beat - for the server to know the job is not dead and refresh the 
'updated_at' property).

Added the suggested shortcut plus another one - on the Progress:
```rust
pub fn update(&self, percent: u8) -> ProgressUpdate {
    set_progress(&self.jid, percent)
}
```
To easily create an update if they see that the percentage last known by the server for this job is out of date, e.g.:
```rust
if progress.percent != Some(100) {
    let upd = progress.update(100);
    assert!(t.lock().unwrap().set_progress(upd).is_ok())
}
```

Reply:
... and removed the redundant `new`

Reply:
EDIT:

After playing with it as an end-user in the e2e tests, I realized that I'd like to have a method `update_percent` which will only update the job completion percentage and will copy over the description, whereas I want to update everything, I could use `update_builder` method returning an instance of `ProgressUpdateBuilder`:
```rust
    pub fn update_percent(&self, percent: u8) -> ProgressUpdate {
        ProgressUpdate::builder(&self.jid)
            .desc(self.desc.clone())
            .percent(percent)
            .build()
    }

    pub fn update_builder(&self) -> ProgressUpdateBuilder {
        ProgressUpdateBuilder::new(&self.jid)
    }
```

Reply:
Isn't the first of those just `self.clone().percent(percent)`?